### ISSUE TITLE
[INCLUDES SOLUTION] Support for Endpoint Routing (ASP.NET Core 3) 

### ISSUE THREAD

**Issue by raz-canva** [AppMetrics/AppMetrics#481]
Currently the route resolvers only work for MvcRouting.

In ASP.NET Core 3, there is now Endpoint Routing.

This means that any endpoints are now supported and their route information is available (not only MVC routes)

---

However, AppMetrics does not resolve these routes so the data is almost useless.

I have a quick fix - but I am not familiar enough with your codebase to know how you want to apply it at a library-level. I have borrowed the logic around `MetricsResourceFilter` `MvcRouteTemplateResolver` to do this.

Hopefully I have provided enough information for you to implement this.

Please let me know if you need any more information.

---

My solution
```csharp
public void Configure(IApplicationBuilder app)
{
app.UseAppMetricsEndpointRoutesResolver();
}
```

```csharp
private static void UseAppMetricsEndpointRoutesResolver(this IApplicationBuilder app)
        {
            app.Use(((context, next) =>
            {
                var metricsCurrentRouteName = "__App.Metrics.CurrentRouteName__";
                var endpointFeature = context.Features[typeof(IEndpointFeature)] as IEndpointFeature;
                if (endpointFeature?.Endpoint is RouteEndpoint endpoint)
                {
                    var method = endpoint.Metadata.GetMetadata<HttpMethodMetadata>()?.HttpMethods
                        ?.FirstOrDefault();
                    var routePattern = endpoint.RoutePattern?.RawText;
                    var templateRouteDetailed = $"{method} {routePattern} {endpoint.DisplayName}";
                    var templateRoute = $"{method} {routePattern}";
                    if (!context.Items.ContainsKey(metricsCurrentRouteName))
                    {
                        context.Items.Add(metricsCurrentRouteName, templateRoute);
                    }
                }

                return next();
            }));
        }
```

**Comment by raz-canva** [comment]
Proposed solution fixes #424 
CC @NinoFloris

**Comment by jvrlic** [comment]
Hi. I use new version (4.0.0.) but routes are not resolved.
There is no metrics output of requests per endpoint with resolved routes.

`public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<Startup> logger)
        {
            app.UseMetricsAllMiddleware();
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UsePathBase("/my/path");

            app.UseErrorLoggingMiddleware();

            app.UseRouting();

            //app.UseAppMetricsEndpointRoutesResolver();

            app.UseAuthentication();
            app.UseAuthorization();
            
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });
            app.UseMetricsEndpoint();
        }`

In ConfigureServices configure metrics:
`var metrics = AppMetrics.CreateDefaultBuilder()
                .OutputMetrics.AsPrometheusPlainText()
                .Configuration.Configure(options =>
                {
                    options.GlobalTags.Add("service", settings.Service_Name);
                    options.AddServerTag(settings.Server_FQDN);
                })
                .Build();

            services.AddMetrics(metrics);

            services.AddMetricsTrackingMiddleware(options =>
            {
                options.IgnoredHttpStatusCodes = settings.MetricsSettings.IgnoredHttpStatusCodes;
            });

            services.AddMetricsEndpoints(options =>
            {
                options.MetricsEndpointOutputFormatter = metrics.OutputMetricsFormatters.OfType<MetricsPrometheusTextOutputFormatter>().First();
            });`
